File: dbassistant/neural/QueryGenerator.py
import os
import numpy as np
import pandas as pd
from keras.models import Sequential
from keras.layers import Dense, Activation
from keras.utils import np_utils
from keras.models import model_from_json
from keras.models import load_model
from tensorflow import keras
from tensorflow.keras import layers
from tensorflow.keras.callbacks import TensorBoard
from dbassistant.analysis.Logging import Logging
from dbassistant.tokens.TokenSeries import TokenSeries
from dbassistant.tokens.Tokenizer import Tokenizer

class QueryGenerator:
    def __init__(self, promptSize, memorySize, optimizer, loss, tokenizer):
        self.promptSize = promptSize
        self.memorySize = memorySize
        self.tokenizer = tokenizer
        self.optimizer = optimizer
        self.loss = loss

        self.model = None
        self.inputLayerSize = memorySize + promptSize
        self.outputLayerSize = tokenizer.labelCount

    def evaluate(self, input:TokenSeries):
        if len(input.tokens) > self.promptSize:
            raise Exception("Prompt size is too large for the network; Input should've been fixed to an acceptable size!")

        Logging.println("    Evaluating tokens: {}...".format(input))

        emptyTokenId = self.tokenizer.handlerOffset(self.tokenizer.emptyTokenHandler)
        outData = []

        endToken = self.tokenizer.handlerOffset(self.tokenizer.endTokenHandler)
        while True:
            output = TokenSeries(outData)
            lastNTokens = output.lastN(self.memorySize, emptyTokenId)
            lastN = lastNTokens.normalizeTokens(self.tokenizer.maxSize)
            prompt = input.normalizeTokens(self.tokenizer.maxSize)
            inData = np.array(lastN + prompt).reshape(1, self.inputLayerSize)

            prediction = self.model.predict(inData)[0]
            maxProbability = np.max(prediction)
            predictedLabel = np.argmax(prediction)
            outData.append(predictedLabel)

            Logging.println("    Prediction: {} (probability: {})".format(predictedLabel, maxProbability))

            if predictedLabel == endToken or len(outData) >= 100:
                break

        result = TokenSeries(outData)
        Logging.println("    Evaluation complete, result: {}".format(result))
        return result

    def loadModel(self, modelFilePath):
        self.model = load_model(modelFilePath)
        Logging.println("Model loaded!")

    def saveModel(self, modelFilePath):
        self.model.save(modelFilePath)
        Logging.println("Model saved!")

    def loadLayers(self, file):
        self.tokenizer.deserialize(file)

    def saveLayers(self, file):
        self.tokenizer.serialize(file)


    def train(self,train_file: str):

        # Load the data from the CSV file, separator is ;
        data = pd.read_csv(train_file, sep=';', header=0)

        # Extract the input values and labels
        # inputs : Array<Array<Float>>
        # labels : Array<Array<Float>>
        inputArrays = data.iloc[:, 0].values
        inputs = []
        for inputArrayString in inputArrays:
            # convert string to array of floats, remove the first and last "
            inputArrayString = inputArrayString[1:-1]
            inputArray = inputArrayString.split(',')
            inputArray = [float(i) for i in inputArray]
            inputs.append(np.array(inputArray))


        labelArrays = data.iloc[:, 1].values
        labels = []
        for labelArrayString in labelArrays:
            # convert string to array of floats
            labelArrayString = labelArrayString[1:-1]
            labelArray = labelArrayString.split(',')
            labelArray = [float(i) for i in labelArray]
            label_count = len(labelArray)
            label = np.argmax(labelArray)
            labels.append(label)

        # Convert inputs and labels to numpy arrays
        row_count = len(inputs)
        input_size = len(inputs[0])
        if input_size != self.inputLayerSize:
            raise Exception("Input size is not equal to the input layer size!")
        inputs = np.array(inputs).reshape(row_count, self.inputLayerSize)
        labels = np.array(labels).reshape(row_count)

        print(f'Inputs shape: {inputs.shape}')
        print(f'Labels shape: {labels.shape}')
        print(f'Output size: {label_count}')

        # Define the model architecture
        self.model = keras.Sequential([
            layers.Dense(150, activation='relu', input_shape=(self.inputLayerSize,)),
            layers.Dense(250, activation='relu'),
            layers.Dense(label_count, activation='softmax')
        ])

        # Compile the model
        self.model.compile(optimizer=self.optimizer,
                    loss=self.loss,
                    metrics=['accuracy'])

        current_path = Logging.currentDir

        # Set up TensorBoard callback
        tensorboard_callback = TensorBoard(log_dir=f'{current_path}/logs/tensorboard', histogram_freq=1, write_graph=True)


        print(f'Training started, you can follow the progress with Tensorboard by running \"tensorboard --logdir {current_path}/logs/tensorboard\"')
        # Train the model with TensorBoard callback
        self.model.fit(inputs, labels, epochs=500, batch_size=64, callbacks=[tensorboard_callback])

        # Evaluate the model
        _, accuracy = self.model.evaluate(inputs, labels)
        print(f'Accuracy: {accuracy * 100}%')

    def __str__(self):
        text = "QueryGenerator Model"
        return text

File: dbassistant/neural/QueryGenerator.ts
import * as tf from '@tensorflow/tfjs-node';
import TokenSeries from '../tokens/TokenSeries'; // Make sure to import TokenSeries and Tokenizer classes
import Tokenizer from '../tokens/Tokenizer'; // Adjust the paths according to your project structure

class QueryGenerator {
    private promptSize: number;
    private memorySize: number;
    private tokenizer: Tokenizer;
    private optimizer: tf.Optimizer;
    private loss: string;
    private model: tf.LayersModel | null = null;
    private inputLayerSize: number;
    private outputLayerSize: number;

    constructor(promptSize: number, memorySize: number, optimizer: tf.Optimizer, loss: string, tokenizer: Tokenizer) {
        this.promptSize = promptSize;
        this.memorySize = memorySize;
        this.tokenizer = tokenizer;
        this.optimizer = optimizer;
        this.loss = loss;

        this.inputLayerSize = memorySize + promptSize;
        this.outputLayerSize = tokenizer.labelCount;
    }

    evaluate(input: TokenSeries): TokenSeries {
        if (input.tokens.length > this.promptSize) {
            throw new Error('Prompt size is too large for the network; Input should\'ve been fixed to an acceptable size!');
        }

        console.log(`    Evaluating tokens: ${input}`);

        const emptyTokenId = this.tokenizer.handlerOffset(this.tokenizer.emptyTokenHandler);
        const outData: number[] = [];

        const endToken = this.tokenizer.handlerOffset(this.tokenizer.endTokenHandler);
        while (true) {
            const output = new TokenSeries(outData);
            const lastNTokens = output.lastN(this.memorySize, emptyTokenId);
            const lastN = lastNTokens.normalizeTokens(this.tokenizer.maxSize);
            const prompt = input.normalizeTokens(this.tokenizer.maxSize);
            const inData = tf.tensor2d([...lastN, ...prompt], [1, this.inputLayerSize]);

            const prediction = this.model!.predict(inData) as tf.Tensor<tf.Rank.R1>;
            const predictionData = prediction.dataSync();
            const maxProbability = Math.max(...predictionData);
            const predictedLabel = predictionData.indexOf(maxProbability);
            outData.push(predictedLabel);

            console.log(`    Prediction: ${predictedLabel} (probability: ${maxProbability})`);

            if (predictedLabel === endToken || outData.length >= 100) {
                break;
            }
        }

        const result = new TokenSeries(outData);
        console.log(`    Evaluation complete, result: ${result}`);
        return result;
    }

    loadModel(modelFilePath: string): void {
        this.model = tf.loadLayersModel(`file://${modelFilePath}`);
        console.log('Model loaded!');
    }

    async saveModel(modelFilePath: string): Promise<void> {
        await this.model!.save(`file://${modelFilePath}`);
        console.log('Model saved!');
    }

    loadLayers(file: string): void {
        this.tokenizer.deserialize(file);
    }

    saveLayers(file: string): void {
        this.tokenizer.serialize(file);
    }

    async train(trainFile: string): Promise<void> {
        // Load the data from the CSV file, separator is ;
        const data = await tf.data.csv(trainFile, { delimiter: ';' });

        // Extract the input values and labels
        const inputArrays = await data.column(0).toArray() as string[];
        const inputs: number[][] = inputArrays.map(inputArrayString => {
            const inputArray = inputArrayString.slice(1, -1).split(',').map(Number);
            return inputArray;
        });

        const labelArrays = await data.column(1).toArray() as string[];
        const labels: number[] = labelArrays.map(labelArrayString => {
            const labelArray = labelArrayString.slice(1, -1).split(',').map(Number);
            const label = labelArray.findIndex(val => val === Math.max(...labelArray));
            return label;
        });

        const rowCount = inputs.length;
        const inputSize = inputs[0].length;
        if (inputSize !== this.inputLayerSize) {
            throw new Error('Input size is not equal to the input layer size!');
        }

        const inputsTensor = tf.tensor2d(inputs);
        const labelsTensor = tf.tensor1d(labels, 'int32');

        console.log(`Inputs shape: ${inputsTensor.shape}`);
        console.log(`Labels shape: ${labelsTensor.shape}`);
        console.log(`Output size: ${this.outputLayerSize}`);

        // Define the model architecture
        this.model = tf.sequential();
        this.model.add(tf.layers.dense({ units: 150, activation: 'relu', inputShape: [this.inputLayerSize] }));
        this.model.add(tf.layers.dense({ units: 250, activation: 'relu' }));
        this.model.add(tf.layers.dense({ units: this.outputLayerSize, activation: 'softmax' }));

        // Compile the model
        this.model.compile({
            optimizer: this.optimizer,
            loss: this.loss,
            metrics: ['accuracy']
        });

        const currentPath = '.'; // Set your desired path here

        // Train the model
        console.log(`Training started, you can follow the progress.`);
        await this.model.fit(inputsTensor, labelsTensor, {
            epochs: 500,
            batchSize: 64,
            callbacks: tf.node.tensorBoard({
                logDir: `${currentPath}/logs/tensorboard`,
                histogramFreq: 1,
                writeGraph: true
            })
        });

        // Evaluate the model
        const evaluation = await this.model.evaluate(inputsTensor, labelsTensor);
        console.log(`Accuracy: ${evaluation[1] * 100}%`);
    }

    toString(): string {
        return 'QueryGenerator Model';
    }
}

export default QueryGenerator;